File: .github/skills/langgraph-project-setup/scripts/init_langgraph_project.js
#!/usr/bin/env node
/**
 * Initialize a new LangGraph JavaScript/TypeScript project.
 *
 * Usage:
 *   node init_langgraph_project.js <project-name> [options]
 *
 * Options:
 *   --path PATH          Output directory (default: current directory)
 *   --pattern PATTERN    Project pattern: simple or multiagent (default: simple)
 *   --graph-name NAME    Graph name for langgraph.json (default: agent)
 *   --typescript         Use TypeScript (default: false)
 */

const fs = require('fs');
const path = require('path');

function parseArgs() {
    const args = process.argv.slice(2);
    const parsed = {
        projectName: args[0],
        path: '.',
        pattern: 'simple',
        graphName: 'agent',
        typescript: false
    };

    for (let i = 1; i < args.length; i++) {
        if (args[i] === '--path' && i + 1 < args.length) {
            parsed.path = args[++i];
        } else if (args[i] === '--pattern' && i + 1 < args.length) {
            parsed.pattern = args[++i];
        } else if (args[i] === '--graph-name' && i + 1 < args.length) {
            parsed.graphName = args[++i];
        } else if (args[i] === '--typescript') {
            parsed.typescript = true;
        }
    }

    if (!parsed.projectName) {
        console.error('Error: Project name is required');
        console.error('Usage: node init_langgraph_project.js <project-name> [options]');
        process.exit(1);
    }

    return parsed;
}

function createDirectoryStructure(projectPath, pattern, useTypescript) {
    const ext = useTypescript ? 'ts' : 'js';
    const srcDir = path.join(projectPath, 'src');

    fs.mkdirSync(srcDir, { recursive: true });

    if (pattern === 'simple') {
        fs.writeFileSync(
            path.join(srcDir, `agent.${ext}`),
            getSimpleAgentTemplate(useTypescript)
        );
    } else {
        const utilsDir = path.join(srcDir, 'utils');
        fs.mkdirSync(utilsDir, { recursive: true });

        fs.writeFileSync(
            path.join(utilsDir, `state.${ext}`),
            getStateTemplate(useTypescript)
        );
        fs.writeFileSync(
            path.join(utilsDir, `nodes.${ext}`),
            getNodesTemplate(useTypescript)
        );
        fs.writeFileSync(
            path.join(utilsDir, `tools.${ext}`),
            getToolsTemplate(useTypescript)
        );
        fs.writeFileSync(
            path.join(srcDir, `agent.${ext}`),
            getMultiagentTemplate(useTypescript)
        );
    }
}

function createLanggraphJson(projectPath, graphName, useTypescript) {
    const ext = useTypescript ? 'ts' : 'js';
    const config = {
        dependencies: ["."],
        graphs: {
            [graphName]: `./src/agent.${ext}:graph`
        },
        env: ".env",
        node_version: "20"
    };

    fs.writeFileSync(
        path.join(projectPath, 'langgraph.json'),
        JSON.stringify(config, null, 2) + '\n'
    );
}

function createEnvFile(projectPath) {
    const content = `# LangSmith Configuration (optional - for tracing and monitoring)
# LANGSMITH_API_KEY=your-api-key-here
# LANGSMITH_TRACING=true
# LANGSMITH_PROJECT=your-project-name

# LLM Provider API Keys (uncomment the ones you need)
# OPENAI_API_KEY=your-openai-api-key-here
# ANTHROPIC_API_KEY=your-anthropic-api-key-here
# GOOGLE_API_KEY=your-google-api-key-here

# Other API Keys
# TAVILY_API_KEY=your-tavily-api-key-here
`;

    fs.writeFileSync(path.join(projectPath, '.env'), content);
}

function createPackageJson(projectPath, projectName, useTypescript) {
    const packageJson = {
        name: projectName,
        version: "0.1.0",
        description: "LangGraph agent application",
        type: "module",
        main: useTypescript ? "dist/agent.js" : "src/agent.js",
        scripts: {
            dev: "langgraph dev",
            ...(useTypescript && {
                build: "tsc",
                typecheck: "tsc --noEmit"
            })
        },
        dependencies: {
            "@langchain/langgraph": "^1.1.0",
            "@langchain/core": "^1.1.0",
            langchain: "^1.1.0"
        },
        devDependencies: {
            ...(useTypescript && {
                typescript: "^5.0.0",
                "@types/node": "^20.0.0"
            })
        }
    };

    fs.writeFileSync(
        path.join(projectPath, 'package.json'),
        JSON.stringify(packageJson, null, 2) + '\n'
    );
}

function createTsConfig(projectPath) {
    const tsconfig = {
        compilerOptions: {
            target: "ES2020",
            module: "ESNext",
            moduleResolution: "node",
            outDir: "./dist",
            rootDir: "./src",
            strict: true,
            esModuleInterop: true,
            skipLibCheck: true,
            forceConsistentCasingInFileNames: true,
            resolveJsonModule: true
        },
        include: ["src/**/*"],
        exclude: ["node_modules", "dist"]
    };

    fs.writeFileSync(
        path.join(projectPath, 'tsconfig.json'),
        JSON.stringify(tsconfig, null, 2) + '\n'
    );
}

function createGitignore(projectPath) {
    const content = `# Dependencies
node_modules/

# Build output
dist/
build/

# Environment
.env
.env.local

# IDE
.vscode/
.idea/
*.swp

# LangGraph
.langgraph/
`;

    fs.writeFileSync(path.join(projectPath, '.gitignore'), content);
}

function getSimpleAgentTemplate(useTypescript) {
    if (useTypescript) {
        return `import { StateGraph, START, END } from "@langchain/langgraph";
import { Annotation, messagesStateReducer } from "@langchain/langgraph";
import { BaseMessage } from "@langchain/core/messages";

// Define agent state
const AgentState = Annotation.Root({
    messages: Annotation<BaseMessage[]>({
        reducer: messagesStateReducer,
    })
});

async function callModel(state: typeof AgentState.State) {
    // TODO: Replace with your LLM provider
    // import { ChatOpenAI } from "@langchain/openai";
    // const model = new ChatOpenAI({ model: "gpt-4" });
    // const response = await model.invoke(state.messages);
    // return { messages: [response] };

    throw new Error("Please configure your LLM provider");
}

// Build graph
const graphBuilder = new StateGraph(AgentState)
    .addNode("call_model", callModel)
    .addEdge(START, "call_model")
    .addEdge("call_model", END);

export const graph = graphBuilder.compile();
`;
    } else {
        return `import { StateGraph, START, END } from "@langchain/langgraph";
import { Annotation, messagesStateReducer } from "@langchain/langgraph";

// Define agent state
const AgentState = Annotation.Root({
    messages: Annotation({
        reducer: messagesStateReducer,
    })
});

async function callModel(state) {
    // TODO: Replace with your LLM provider
    // import { ChatOpenAI } from "@langchain/openai";
    // const model = new ChatOpenAI({ model: "gpt-4" });
    // const response = await model.invoke(state.messages);
    // return { messages: [response] };

    throw new Error("Please configure your LLM provider");
}

// Build graph
const graphBuilder = new StateGraph(AgentState)
    .addNode("call_model", callModel)
    .addEdge(START, "call_model")
    .addEdge("call_model", END);

export const graph = graphBuilder.compile();
`;
    }
}

function getMultiagentTemplate(useTypescript) {
    const imports = useTypescript
        ? `import { StateGraph, START, END } from "@langchain/langgraph";
import { AgentState } from "./utils/state.js";
import { supervisor, worker1, worker2 } from "./utils/nodes.js";`
        : `import { StateGraph, START, END } from "@langchain/langgraph";
import { AgentState } from "./utils/state.js";
import { supervisor, worker1, worker2 } from "./utils/nodes.js";`;

    return `${imports}

function buildGraph() {
    const graphBuilder = new StateGraph(AgentState)
        .addNode("supervisor", supervisor)
        .addNode("worker_1", worker1)
        .addNode("worker_2", worker2)
        .addEdge(START, "supervisor")
        .addConditionalEdges(
            "supervisor",
            (state) => state.nextAgent || "END",
            {
                worker_1: "worker_1",
                worker_2: "worker_2",
                END: END
            }
        )
        .addEdge("worker_1", "supervisor")
        .addEdge("worker_2", "supervisor");

    return graphBuilder.compile();
}

export const graph = buildGraph();
`;
}

function getStateTemplate(useTypescript) {
    if (useTypescript) {
        return `import { Annotation, messagesStateReducer } from "@langchain/langgraph";
import { BaseMessage } from "@langchain/core/messages";

export const AgentState = Annotation.Root({
    messages: Annotation<BaseMessage[]>({
        reducer: messagesStateReducer,
    }),
    nextAgent: Annotation<"worker_1" | "worker_2" | "END">({
        reducer: (_, next) => next,
        default: () => "END"
    })
});
`;
    } else {
        return `import { Annotation, messagesStateReducer } from "@langchain/langgraph";

export const AgentState = Annotation.Root({
    messages: Annotation({
        reducer: messagesStateReducer,
    }),
    nextAgent: Annotation({
        reducer: (_, next) => next,
        default: () => "END"
    })
});
`;
    }
}

function getNodesTemplate(useTypescript) {
    const typeAnnotations = useTypescript ? ': typeof AgentState.State' : '';

    return `import { AIMessage } from "@langchain/core/messages";
${useTypescript ? 'import { AgentState } from "./state.js";' : ''}

export async function supervisor(state${typeAnnotations}) {
    const messages = state.messages;

    // TODO: Implement supervisor logic
    if (messages.length > 2) {
        return { nextAgent: "END" };
    }

    return { nextAgent: "worker_1" };
}

export async function worker1(state${typeAnnotations}) {
    // TODO: Implement worker 1 logic
    const response = new AIMessage({ content: "Worker 1 response" });
    return { messages: [response] };
}

export async function worker2(state${typeAnnotations}) {
    // TODO: Implement worker 2 logic
    const response = new AIMessage({ content: "Worker 2 response" });
    return { messages: [response] };
}
`;
}

function getToolsTemplate(useTypescript) {
    const typeAnnotations = useTypescript ? ': string' : '';

    return `import { tool } from "@langchain/core/tools";
import { z } from "zod";

export const exampleTool = tool(
    async (input${typeAnnotations}) => {
        return \`Tool received: \${input.query}\`;
    },
    {
        name: "example_tool",
        description: "Example tool that returns the input query",
        schema: z.object({
            query: z.string().describe("The input query string"),
        }),
    }
);

export const tools = [exampleTool];
`;
}

function main() {
    const args = parseArgs();
    const basePath = path.resolve(args.path);
    const projectPath = path.join(basePath, args.projectName);

    if (fs.existsSync(projectPath)) {
        console.error(`❌ Error: Directory ${projectPath} already exists`);
        process.exit(1);
    }

    console.log(`🚀 Initializing LangGraph project: ${args.projectName}`);
    console.log(`   Pattern: ${args.pattern}`);
    console.log(`   Language: ${args.typescript ? 'TypeScript' : 'JavaScript'}`);
    console.log(`   Location: ${projectPath}`);
    console.log();

    fs.mkdirSync(projectPath, { recursive: true });

    createDirectoryStructure(projectPath, args.pattern, args.typescript);
    createLanggraphJson(projectPath, args.graphName, args.typescript);
    createEnvFile(projectPath);
    createPackageJson(projectPath, args.projectName, args.typescript);
    if (args.typescript) {
        createTsConfig(projectPath);
    }
    createGitignore(projectPath);

    console.log('✅ Created project structure');
    console.log('✅ Created langgraph.json');
    console.log('✅ Created .env template');
    console.log('✅ Created package.json');
    if (args.typescript) {
        console.log('✅ Created tsconfig.json');
    }
    console.log('✅ Created .gitignore');
    console.log();
    console.log('📦 Next steps:');
    console.log(`   1. cd ${args.projectName}`);
    console.log('   2. Install dependencies: npm install');
    console.log('   3. Configure .env with your API keys');
    console.log('   4. Start development server: npm run dev');
    console.log();
    console.log('🎯 Happy building!');
}

main();

File: .github/skills/langgraph-project-setup/scripts/init_langgraph_project.py
#!/usr/bin/env python3
"""
Initialize a new LangGraph project with proper structure and configuration.

Usage:
    uv run init_langgraph_project.py <project-name> [options]

Fallback:
    python3 init_langgraph_project.py <project-name> [options]

Options:
    --path PATH          Output directory (default: current directory)
    --pattern PATTERN    Project pattern: simple or multiagent (default: simple)
    --graph-name NAME    Graph name for langgraph.json (default: agent)
    --python-version VER Python version: 3.11, 3.12, or 3.13 (default: 3.11)
"""

import argparse
import json
import sys
from pathlib import Path
from typing import Literal


def create_directory_structure(
    project_path: Path,
    pattern: Literal["simple", "multiagent"]
) -> None:
    """Create the LangGraph project directory structure."""

    # Create main package directory (use underscore for Python package naming)
    package_name = project_path.name.replace("-", "_")
    package_dir = project_path / package_name
    package_dir.mkdir(parents=True, exist_ok=True)

    # Create __init__.py for main package
    (package_dir / "__init__.py").write_text("")

    if pattern == "simple":
        # Simple pattern: just agent.py at package root
        (package_dir / "agent.py").write_text(get_simple_agent_template())
    else:
        # Multi-agent pattern: utils directory with separated concerns
        utils_dir = package_dir / "utils"
        utils_dir.mkdir(exist_ok=True)

        # Create __init__.py for utils
        (utils_dir / "__init__.py").write_text("")

        # Create modular files
        (utils_dir / "state.py").write_text(get_state_template())
        (utils_dir / "nodes.py").write_text(get_nodes_template())
        (utils_dir / "tools.py").write_text(get_tools_template())
        (package_dir / "agent.py").write_text(get_multiagent_template(package_name))


def create_langgraph_json(
    project_path: Path,
    graph_name: str,
    python_version: str
) -> None:
    """Create langgraph.json configuration file."""

    package_name = project_path.name.replace("-", "_")

    config = {
        "dependencies": ["."],
        "graphs": {
            graph_name: f"./{package_name}/agent.py:graph"
        },
        "env": ".env",
        "python_version": python_version
    }

    config_path = project_path / "langgraph.json"
    config_path.write_text(json.dumps(config, indent=2) + "\n")


def create_env_file(project_path: Path) -> None:
    """Create .env template file with common environment variables."""

    env_content = """# LangSmith Configuration (optional - for tracing and monitoring)
# LANGSMITH_API_KEY=your-api-key-here
# LANGSMITH_TRACING=true
# LANGSMITH_PROJECT=your-project-name

# LLM Provider API Keys (uncomment the ones you need)
# OPENAI_API_KEY=your-openai-api-key-here
# ANTHROPIC_API_KEY=your-anthropic-api-key-here
# GOOGLE_API_KEY=your-google-api-key-here

# AWS Bedrock (if using AWS)
# AWS_ACCESS_KEY_ID=your-aws-access-key-id
# AWS_SECRET_ACCESS_KEY=your-aws-secret-key
# AWS_DEFAULT_REGION=us-east-1

# Other API Keys
# TAVILY_API_KEY=your-tavily-api-key-here
"""

    env_path = project_path / ".env"
    env_path.write_text(env_content)


def create_pyproject_toml(project_path: Path) -> None:
    """Create pyproject.toml with dependencies."""

    package_name = project_path.name.replace("-", "_")

    content = f"""[project]
name = "{package_name}"
version = "0.1.0"
description = "LangGraph agent application"
requires-python = ">=3.10"
dependencies = [
    "langgraph>=1.1.0",
    "langchain-core>=1.1.0",
    "langchain>=1.1.0",
]

[project.optional-dependencies]
openai = ["langchain-openai>=1.1.0"]
anthropic = ["langchain-anthropic>=1.1.0"]
google = ["langchain-google-genai>=4.0.0"]
dev = [
    "langgraph-cli[inmem]>=0.4.0",
    "pytest>=7.0.0",
]

[build-system]
requires = ["setuptools>=61.0"]
build-backend = "setuptools.build_meta"
"""

    toml_path = project_path / "pyproject.toml"
    toml_path.write_text(content)


def create_gitignore(project_path: Path) -> None:
    """Create .gitignore file."""

    content = """# Python
__pycache__/
*.py[cod]
*$py.class
*.so
.Python
build/
develop-eggs/
dist/
downloads/
eggs/
.eggs/
lib/
lib64/
parts/
sdist/
var/
wheels/
*.egg-info/
.installed.cfg
*.egg

# Virtual environments
venv/
env/
ENV/
.venv

# Environment variables
.env
.env.local

# IDE
.vscode/
.idea/
*.swp
*.swo
*~

# LangGraph
.langgraph/
"""

    gitignore_path = project_path / ".gitignore"
    gitignore_path.write_text(content)


def get_simple_agent_template() -> str:
    """Return template for simple agent.py."""
    return '''"""Simple LangGraph agent."""

from typing import Annotated
from typing_extensions import TypedDict

from langgraph.graph import StateGraph, START, END
from langgraph.graph.message import add_messages
from langchain_core.messages import BaseMessage


class State(TypedDict):
    """Agent state."""
    messages: Annotated[list[BaseMessage], add_messages]


def call_model(state: State) -> dict:
    """Call the language model."""
    # TODO: Replace with your LLM provider
    # from langchain_openai import ChatOpenAI
    # model = ChatOpenAI(model="gpt-4")
    # response = model.invoke(state["messages"])
    # return {"messages": [response]}

    raise NotImplementedError("Please configure your LLM provider")


# Build graph
graph_builder = StateGraph(State)
graph_builder.add_node("call_model", call_model)
graph_builder.add_edge(START, "call_model")
graph_builder.add_edge("call_model", END)

graph = graph_builder.compile()
'''


def get_multiagent_template(package_name: str) -> str:
    """Return template for multi-agent agent.py."""
    return f'''"""Multi-agent LangGraph application."""

from langgraph.graph import StateGraph, START, END
from {package_name}.utils.state import AgentState
from {package_name}.utils.nodes import supervisor, worker_1, worker_2


def build_graph() -> StateGraph:
    """Build the multi-agent graph."""

    graph_builder = StateGraph(AgentState)

    # Add nodes
    graph_builder.add_node("supervisor", supervisor)
    graph_builder.add_node("worker_1", worker_1)
    graph_builder.add_node("worker_2", worker_2)

    # Add edges
    graph_builder.add_edge(START, "supervisor")
    graph_builder.add_conditional_edges(
        "supervisor",
        lambda state: state.get("next_agent", "END"),
        {{
            "worker_1": "worker_1",
            "worker_2": "worker_2",
            "END": END
        }}
    )
    graph_builder.add_edge("worker_1", "supervisor")
    graph_builder.add_edge("worker_2", "supervisor")

    return graph_builder.compile()


graph = build_graph()
'''


def get_state_template() -> str:
    """Return template for state.py."""
    return '''"""State definitions for the agent."""

from typing import Annotated, Literal
from typing_extensions import TypedDict

from langgraph.graph.message import add_messages
from langchain_core.messages import BaseMessage


class AgentState(TypedDict):
    """State for multi-agent system."""
    messages: Annotated[list[BaseMessage], add_messages]
    next_agent: Literal["worker_1", "worker_2", "END"]
'''


def get_nodes_template() -> str:
    """Return template for nodes.py."""
    return '''"""Node functions for the agent."""

from langchain_core.messages import HumanMessage, AIMessage
from .state import AgentState


def supervisor(state: AgentState) -> dict:
    """Supervisor node that routes to workers."""
    messages = state["messages"]

    # TODO: Implement supervisor logic
    # This is a placeholder that routes to worker_1
    # In practice, you might use an LLM to decide routing

    if len(messages) > 2:
        return {"next_agent": "END"}

    return {"next_agent": "worker_1"}


def worker_1(state: AgentState) -> dict:
    """First worker node."""
    # TODO: Implement worker 1 logic
    response = AIMessage(content="Worker 1 response")
    return {"messages": [response]}


def worker_2(state: AgentState) -> dict:
    """Second worker node."""
    # TODO: Implement worker 2 logic
    response = AIMessage(content="Worker 2 response")
    return {"messages": [response]}
'''


def get_tools_template() -> str:
    """Return template for tools.py."""
    return '''"""Tools for the agent."""

from langchain_core.tools import tool


@tool
def example_tool(query: str) -> str:
    """Example tool that returns the input query.

    Args:
        query: The input query string

    Returns:
        The input query as output
    """
    return f"Tool received: {query}"


# List of all tools
tools = [example_tool]
'''


def main():
    parser = argparse.ArgumentParser(
        description="Initialize a new LangGraph project"
    )
    parser.add_argument(
        "project_name",
        help="Name of the project (e.g., my-agent)"
    )
    parser.add_argument(
        "--path",
        default=".",
        help="Output directory (default: current directory)"
    )
    parser.add_argument(
        "--pattern",
        choices=["simple", "multiagent"],
        default="simple",
        help="Project pattern (default: simple)"
    )
    parser.add_argument(
        "--graph-name",
        default="agent",
        help="Graph name for langgraph.json (default: agent)"
    )
    parser.add_argument(
        "--python-version",
        choices=["3.11", "3.12", "3.13"],
        default="3.11",
        help="Python version (default: 3.11)"
    )

    args = parser.parse_args()

    # Create project directory
    base_path = Path(args.path).resolve()
    project_path = base_path / args.project_name

    if project_path.exists():
        print(f"❌ Error: Directory {project_path} already exists")
        sys.exit(1)

    print(f"🚀 Initializing LangGraph project: {args.project_name}")
    print(f"   Pattern: {args.pattern}")
    print(f"   Location: {project_path}")
    print()

    # Create all files and directories
    create_directory_structure(project_path, args.pattern)
    create_langgraph_json(project_path, args.graph_name, args.python_version)
    create_env_file(project_path)
    create_pyproject_toml(project_path)
    create_gitignore(project_path)

    print("✅ Created project structure")
    print("✅ Created langgraph.json")
    print("✅ Created .env template")
    print("✅ Created pyproject.toml")
    print("✅ Created .gitignore")
    print()
    print("📦 Next steps:")
    print(f"   1. cd {args.project_name}")
    print(f"   2. Create a virtual environment: uv venv --python {args.python_version}")
    print("      Fallback: python3 -m venv venv")
    print("   3. Activate it: source .venv/bin/activate")
    print("      Fallback: source venv/bin/activate")
    print("   4. Install dependencies: uv pip install -e '.[dev]'")
    print("      Fallback: pip install -e '.[dev]'")
    print("   5. Configure .env with your API keys")
    print("   6. Start development server: langgraph dev")
    print()
    print("🎯 Happy building!")


if __name__ == "__main__":
    main()
